server/pairing/client.go

```
@@ -232,169 +238,236 @@ func (c *Client) receiveInstallationData() error {
 		return err
 	}
-	signal.SendLocalPairingEvent(Event{Type: EventProcessSuccess, Action: ActionPairingAccount})
+	signal.SendLocalPairingEvent(Event{Type: EventProcessSuccess, Action: ActionSyncDevice})
 	return nil
 }
 
-func (c *Client) getChallenge() error {
-	c.baseAddress.Path = pairingChallenge
-	resp, err := c.Get(c.baseAddress.String())
+func (c *ReceiverClient) sendInstallationData() error {
+	err := c.installationReceiver.Mount()
 	if err != nil {
 		return err
 	}
 
-	if resp.StatusCode != http.StatusOK {
-		return fmt.Errorf("[client] status not ok when getting challenge, received '%s'", resp.Status)
+	c.baseAddress.Path = pairingReceiveInstallation
+	req, err := http.NewRequest(http.MethodPost, c.baseAddress.String(), bytes.NewBuffer(c.installationReceiver.ToSend()))
+	if err != nil {
+		return err
 	}
+	req.Header.Set("Content-Type", "application/octet-stream")
 
-	c.serverChallenge, err = ioutil.ReadAll(resp.Body)
-	return err
-}
-
-func StartUpPairingClient(backend *api.GethStatusBackend, cs, configJSON string) error {
-	c, err := setupClient(backend, cs, configJSON)
+	err = c.doChallenge(req)
 	if err != nil {
+		signal.SendLocalPairingEvent(Event{Type: EventTransferError, Error: err.Error(), Action: ActionPairingInstallation})
 		return err
 	}
-	err = c.PairAccount()
+
+	resp, err := c.Do(req)
 	if err != nil {
+		signal.SendLocalPairingEvent(Event{Type: EventTransferError, Error: err.Error(), Action: ActionPairingInstallation})
 		return err
 	}
-	err = c.PairSyncDevice()
-	if err != nil {
+
+	if resp.StatusCode != http.StatusOK {
+		err = fmt.Errorf("[client] status not okay when sending installation data, status: %s", resp.Status)
+		signal.SendLocalPairingEvent(Event{Type: EventTransferError, Error: err.Error(), Action: ActionPairingInstallation})
 		return err
 	}
-	return c.PairInstallation()
+
+	signal.SendLocalPairingEvent(Event{Type: EventTransferSuccess, Action: ActionPairingInstallation})
```

Just a question. I see under the hood `SendLocalPairingEvent` uses `signals.go#send`, which can fail due to marshalling, but the error is not returned (only logged). Is this because of the fallback to the C implementation? What do you think is the reason the `send` function doesn't return an error?

---

Ah, you've really been digging into this code.

Ultimately signals are treated as outside the main critical path, so if a signal fails that failure should not interrupt the current process flow.

Additionally `json.Marshal()` generally only fails if you make a mistake in the code, and only if you are trying to marshal abstract things. There are 2 possible errors from this package `UnsupportedTypeError` and `UnsupportedValueError`, this means you've passed in something too abstract to represent as a JSON object.

You can see in [the source code for the go `json` package](https://github.com/golang/go/blob/master/src/encoding/json/encode.go#L426-L428) that if a type doesn't have a specific handler an `unsupportedTypeEncoder` will be returned.

Example:

```go
j, err := json.Marshal(make(chan struct{}))
if _, ok := err.(*json.UnsupportedTypeError); ok {
	fmt.Println("json error : json.UnsupportedTypeError")
}
fmt.Println("marshalled json : '" + string(j) + "'")

j, err = json.Marshal(func() {})
if _, ok := err.(*json.UnsupportedTypeError); ok {
	fmt.Println("json error : json.UnsupportedTypeError")
}
fmt.Println("marshalled json : '" + string(j) + "'")

j, err = json.Marshal(complex128(1))
if _, ok := err.(*json.UnsupportedTypeError); ok {
	fmt.Println("json error : json.UnsupportedTypeError")
}
fmt.Println("marshalled json : '" + string(j) + "'")

j, err = json.Marshal(math.Inf(1))
if _, ok := err.(*json.UnsupportedValueError); ok {
	fmt.Println("json error : json.UnsupportedValueError")
}
fmt.Println("marshalled json : '" + string(j) + "'")
```

Output:

```text
json error : json.UnsupportedTypeError
marshalled json : ''
json error : json.UnsupportedTypeError
marshalled json : ''
json error : json.UnsupportedTypeError
marshalled json : ''
json error : json.UnsupportedValueError
marshalled json : ''

Program exited.
```

See : https://go.dev/play/p/1dC31lpuB07

So you really would be trying to break the marshaller if you get an error from `json.Marshal()`. But ordinarily the error would be handled, as the error would constitute a critical failure.

With a signal failure, the worst that happens is that the signal is not sent, which although inconvenient isn't a processes halting

---

Now it all makes sense to me, we're good after all :) Thank you for the detailed explanation!